### PR TITLE
Feature/api with generics

### PR DESCRIPTION
**Pull request by TobTheRock**

### CODE REVIEW

File: src/sender.rs [hunk 1]
```diff
@@ -14,35 +14,47 @@ use crate::{
...
             cipher_suite,
             secret: None,
             buffer: Default::default(),
         }
     }
 
-    pub fn encrypt(&mut self, unencrypted_payload: &[u8], skip: usize) -> Result<&[u8]> {
+    pub fn encrypt<Payload>(&mut self, unencrypted_payload: &Payload, skip: usize) -> Result<&[u8]>
+    where
+        Payload: AsRef<[u8]> + ?Sized,
```
Comment (hoodie) [inline src/sender.rs:54]: are you explicitly using two different terms here? `Payload` vs `Frame`?

File: src/sender.rs [hunk 2]
```diff
@@ -88,8 +100,12 @@ impl Sender {
         }
     }
 
-    pub fn set_encryption_key(&mut self, key_material: &[u8]) -> Result<()> {
-        self.secret = Some(KeyMaterial(key_material).expand_as_secret(&self.cipher_suite)?);
+    pub fn set_encryption_key<M>(&mut self, key_material: &M) -> Result<()>
+    where
+        M: AsRef<[u8]> + ?Sized,
```
Comment (hoodie) [inline src/sender.rs:105]: `K` for key?

File: src/sender.rs [hunk 3]
```diff
@@ -14,35 +14,47 @@ use crate::{
...
     secret: Option<Secret>,
     buffer: Vec<u8>,
 }
 
 impl Sender {
-    pub fn new(sender_id: u64) -> Sender {
-        Self::with_cipher_suite(sender_id, CipherSuiteVariant::AesGcm256Sha512)
+    pub fn new<K>(key_id: K) -> Sender
+    where
+        K: Into<KeyId>,
```
Comment (hoodie) [inline src/sender.rs:26]: `I` for Id?

File: src/sender.rs (hunk 2 above)
Comment (hoodie) [reply src/sender.rs:105]: actually, M is fine, you wanna distinguish the "id" from the "material" right?

File: src/sender.rs (hunk 1 above)
Comment (hoodie) [reply src/sender.rs:54]: actually while we're at it, [the standard in this place calls it "plaintext"](https://sframe-wg.github.io/sframe/draft-ietf-sframe-enc.html#name-encryption), which might be a bit more clear

Comment (TobTheRock) [reply src/sender.rs]: I like it :+1: For decrypting I have choosen ÈncryptedFrame` in contrary to `Sframe`, which sounded weird imho

File: src/sender.rs (hunk 2 above)
Comment (TobTheRock) [reply src/sender.rs]: I've named the parameters more specificly

File: src/sender.rs (hunk 3 above)
Comment (TobTheRock) [reply src/sender.rs:26]: Using `Id`now

Comment (hoodie) [reply src/sender.rs:26]: a matter of taste really. I personally like single character type-parameters, as you can more clearly distinguish them from actual types.

**Review by hoodie — APPROVED** [review]

File: src/sender.rs (hunk 3 above)
Comment (TobTheRock) [reply src/sender.rs:26]: Yes, a good topic for endless discussions :D

**Pull request merged**